nav2_costmap_2d/src/clear_costmap_service.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <memory>
+
+#include "nav2_costmap_2d/clear_costmap_service.hpp"
+#include "nav2_costmap_2d/costmap_2d_ros.hpp"
+
+namespace nav2_costmap_2d
+{
+
+using std::vector;
+using std::string;
+using std::shared_ptr;
+using std::any_of;
+using ClearExceptRegion = nav2_msgs::srv::ClearCostmapExceptRegion;
+
+ClearCostmapService::ClearCostmapService(rclcpp::Node::SharedPtr & node, Costmap2DROS * costmap)
+: node_(node), costmap_(costmap)
+{
+  node_->get_parameter_or_set("clearable_layers", clearable_layers_, {"obstacle_layer"});
+
+  server_ = node_->create_service<ClearExceptRegion>("clear_except_" + costmap->getName(),
+      std::bind(&ClearCostmapService::clearExceptRegionCallback, this,
+      std::placeholders::_1, std::placeholders::_2, std::placeholders::_3));
+}
+
+void ClearCostmapService::clearExceptRegionCallback(
+  const shared_ptr<rmw_request_id_t>/*request_header*/,
+  const shared_ptr<ClearExceptRegion::Request> request,
+  const shared_ptr<ClearExceptRegion::Response>/*response*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "Received request to clear " + costmap_->getName());
+
+  if (costmap_ == nullptr) {
+    RCLCPP_ERROR(node_->get_logger(), "Costmap is undefined. Doing nothing.");
+    return;
+  }
+
+  RCLCPP_INFO(node_->get_logger(), "Will proceed with clearing the costmap");
+
+  clearExceptRegion(request->reset_distance);
+}
+
+void ClearCostmapService::clearExceptRegion(const double reset_distance)
+{
+  double x, y;
+
+  if (!getPosition(x, y)) {
+    RCLCPP_ERROR(node_->get_logger(), "Cannot clear map because robot pose cannot be retrieved.");
+    return;
+  }
+
+  auto layers = costmap_->getLayeredCostmap()->getPlugins();
+
+  for (auto & layer : *layers) {
+    if (isClearable(getLayerName(*layer))) {
+      auto costmap_layer = std::static_pointer_cast<CostmapLayer>(layer);
+      clearLayerExceptRegion(costmap_layer, x, y, reset_distance);
+    }
+  }
+}
+
+bool ClearCostmapService::isClearable(const string & layer_name) const
+{
+  return any_of(begin(clearable_layers_), end(clearable_layers_),
+           [&layer_name](auto l) {return l == layer_name;});
```

Comment:
I'd suggest
```
auto it = find(begin(clearable_layers_), end(clearable_layers_), layer_name);
return it != end(clearable_layers_);
```
It's a bit more direct and probably easier to parse

Reply:
It's debatable which is better to use for this case, if `find`, `count` or `any_of`. 

I like `find` and `count` because we avoid using a lamba, however, I feel `any_of` express the intention more directly.

I'll change it to:
```cpp
return count(begin(clearable_layers_), end(clearable_layers_), layer_name) != 0;
```

Reply:
I'd say `find` seems to be the default when you want to determine if a container contains a value, but I'll admit comparing to 0 is a touch prettier than `end(clearable_layers_)`

Find is likely to be more efficient on a large container since it can stop at the first instance instead of iterating to the end, however, in this case the list is likely to be very small so it doesn't really matter.